Make client.Context implement gRPC ClientConn

aaronc opened the pull request:
ref: #5921 

This PR adds:
* an implementation of gRPC `ClientConn` on `client.Context` which proxies requests to the ABCI query endpoint
* a gRPC `TestService` to simapp
* an integration test to verify that `client.Context` can make gRPC queries over ABCI

---

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

alexanderbez commented on client/grpc_query.go:
```diff
@@ -0,0 +1,43 @@
+var protoCodec = encoding.GetCodec(proto.Name)
+
+// Invoke implements the grpc ClientConn.Invoke method
+func (c cliQueryConn) Invoke(_ gocontext.Context, method string, args, reply interface{}, _ ...grpc.CallOption) error {
```
Out of curiosity, `cliQueryConn` is just a wrapper around `ctx`, why not simply have `Context` implement `gogogrpc.ClientConn`?

aaronc replied:
Oh l think that could work! 👍

aaronc commented:
I updated this to just use `client.Context`. How does it look now @alexanderbez? Not sure what to do about code coverage because I added an integration test but that isn't reflected in coverage.

alessio commented on codec/testdata/test_service.go:
```diff
@@ -0,0 +1,19 @@
+package testdata
```
Why is package named `testdata`? It could be very confusing, let alone that it does not seem to meet Golang's conventions on `testdata` directories [1]

[1] https://golang.org/pkg/cmd/go/internal/test/

aaronc replied:
Can you suggest a better name?

alessio replied:
`testservice`?

aaronc replied:
Well it's not just the service, it's all out test protobuf data types, thus `testdata`

alessio replied:
`testdatatypes`?

Anyway, I see. The `go` tool ignores `testdata` directories, so as long as this does not represent a problem, it should be fine.

aaronc replied:
Maybe we keep it for now and if it's a problem we can change later? So far it's been working.

fedekunze commented on client/grpc_query.go:
```diff
@@ -0,0 +1,33 @@
+	if err != nil {
+		return err
+	}
+	resBz, _, err := ctx.QueryWithData(method, reqBz)
```
nit, should every reply be wrapped with the response height? cc @alexanderbez

alexanderbez replied:
I'm not sure how that would work here. I believe the response `resBz` already has that info? Correct me if I'm wrong @aaronc
